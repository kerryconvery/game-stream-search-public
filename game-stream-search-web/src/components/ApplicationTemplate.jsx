import React from 'react';
import { node } from 'prop-types';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';

const ApplicationTemplate = ({ alertRenderer, toastRenderer, children }) => {
  return (
    <Container maxWidth='xl' disableGutters>
      <Grid container direction="column" alignItems="stretch">
        {alertRenderer}
        {toastRenderer}
        {children}
      </Grid>
    </Container>
  )
}

ApplicationTemplate.propTypes = {
  alertRenderer: node.isRequired,
  toastRenderer: node.isRequired,
  children: node.isRequired,
}

export default ApplicationTemplate;
