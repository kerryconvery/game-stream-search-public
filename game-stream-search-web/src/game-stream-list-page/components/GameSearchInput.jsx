import React, { useState } from 'react';
import { func } from 'prop-types';
import _has from 'lodash/has';
import _trim from 'lodash/trim';
import { makeStyles } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import InputBase from '@material-ui/core/InputBase';
import SearchIcon from '@material-ui/icons/Search';
import IconButton from '@material-ui/core/IconButton';

const useStyles = makeStyles((theme) => ({
  root: {
    padding: '2px 4px',
    display: 'flex',
    alignItems: 'center',
    width: '500px',
    margin: 'auto',
  },
  input: {
    marginLeft: theme.spacing(1),
    flex: 1,
    fontFamily: 'Helvetica'
  },
  iconButton: {
    padding: 1,
  },
}));

const GameStreamInput = ({ onGameChange }) => {
  const [ input, setInput ] = useState('');
  
  const classes = useStyles();

  const onInputChange = (e) => {
    if (_has(e, 'target.value')) {
      setInput(e.target.value);
    }
  }
  const onKeyPress = (e) => {
    if(e.keyCode === 13 && _has(e, 'target.value')) {
      onClick(e.target.value)
    }
  }

  const onClick = (value) => {
    onGameChange(_trim(value) !== '' ? value : null);
  }

  return (
    <Paper className={classes.root} variant='outlined'>
      <InputBase
        className={classes.input}
        variant='outlined'
        placeholder="Search"
        inputProps={{ 'aria-label': 'search' }}
        onChange={onInputChange}
        onKeyDown={onKeyPress}
      />
      <IconButton className={classes.iconButton} aria-label="search" onClick={() => onGameChange(input)}>
        <SearchIcon />
      </IconButton>
    </Paper>
  );
};

GameStreamInput.propTypes = {
  onGameChange: func.isRequired,
}

export default GameStreamInput;
