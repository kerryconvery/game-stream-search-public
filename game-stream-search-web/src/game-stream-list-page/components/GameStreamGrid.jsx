import React from 'react';
import { node, string, number } from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import GridList from '@material-ui/core/GridList';
import GridListTile from '@material-ui/core/GridListTile';
import GridListTileBar from '@material-ui/core/GridListTileBar';
import Skeleton from '@material-ui/lab/Skeleton';
import Grid from '@material-ui/core/Grid';
import IconButton from '@material-ui/core/IconButton';
import PlayArrowIcon from '@material-ui/icons/PlayArrow';
import PlayCircleOutlineIcon from '@material-ui/icons/PlayCircleOutline';
import Avatar from '@material-ui/core/Avatar';
import Tooltip from '@material-ui/core/Tooltip';

const useStreamTileStyles = makeStyles(() => ({
  root: {
    'font-family': 'Helvetica',
    color: '#231E1D',
  },
  streamTitle: {
    display: '-webkit-box',
    '-webkit-box-orient': 'vertical',
    '-webkit-line-clamp': '2',
    fontSize: '16px',
    textOverflow: 'ellipsis',
    overflow: 'hidden',
    whiteSpace: 'normal',
    marginTop: 0,
    marginBottom: '0.25rem',
  },
  link: {
    textDecoration: 'none',
    color: 'inherit',
    '&:hover > button': {
      opacity: 0.7
    }
  },
  detailsContainer: {
    display: 'flex',
    flexDirection: 'row',
    paddingTop: '1rem',
  },
  streamDetails: {
    paddingLeft: '0.5rem',
  },
  streamSubDetails: {
    color: '#606060',
    fontSize: '14px',
  },
  playButton: {
    color: 'white',
    opacity: 0,
    position: 'absolute',
    left: 0, 
    right: 0,
    top: '10%',
    marginLeft: 'auto',
    marginRight: 'auto', 
    width: '100px', /* Need a specific value to work */
    transition: 'opacity 0.5s ease',
  },
  playButtonIcon: {
    fontSize: 100,
    color: 'white',
  },
}));


const StreamTile = ({
  streamTitle,
  streamThumbnailUrl,
  streamUrl,
  streamerName,
  streamerAvatarUrl,
  platformName,
  views }) => {

  const classes = useStreamTileStyles();

  return (
    <div className={classes.root}>
      <a href={streamUrl} target='_blank' className={classes.link}>
        <IconButton size="medium" className={classes.playButton}>
          <PlayCircleOutlineIcon className={classes.playButtonIcon} />
        </IconButton >
        <img src={streamThumbnailUrl} width={320} height={180} />
        <div className={classes.detailsContainer}>
          <Avatar src={streamerAvatarUrl} />
          <div className={classes.streamDetails}>
            <Tooltip title={streamTitle}>
              <h3 className={classes.streamTitle}>{streamTitle}</h3>
            </Tooltip>
            <div className={classes.streamSubDetails}>
              <div>{streamerName}</div>
              <span>{platformName}</span>
              <span> â€¢ </span>
              <span>{`${views} viewers`}</span>
            </div>
          </div>
        </div>
      </a>
    </div>
  )
}

StreamTile.propTypes = {
  streamTitle: string.isRequired,
  streamerName: string.isRequired,
  streamThumbnailUrl: string.isRequired,
  streamUrl: string.isRequired,
  streamerAvatarUrl: string.isRequired,
  platformName: string.isRequired,
  views: number.isRequired,
}

const useLoadingTileStyles = makeStyles(() => ({
  detailsContainer: {
    display: 'flex',
    flexDirection: 'row',
    paddingTop: '1rem',
    width: '100%',
    height: '100%',
  },
  details: {
    paddingLeft: '0.5rem',
    width: '80%',
    height: '80%',
  },
  subDetails: {
    width: '50%',
  },
}));

const LoadingTile = () => {
  const classes = useLoadingTileStyles();

  return (
    <>
      <Skeleton variant='rect' height='60%' animation='wave' />
      <div className={classes.detailsContainer}>
        <Skeleton variant='circle' width={50} height={50} animation='wave' />
        <div className={classes.details}>
          <Skeleton variant='text' animation='wave' />
          <div className={classes.subDetails}>
            <Skeleton variant='text' animation='wave' />
            <Skeleton variant='text' animation='wave' />
          </div>
        </div>
      </div>
    </>
  )
}

export const getLoadingTiles = (size) => {
  const loadingTiles = [];

  for (let index = 0; index < size; index++) {
    loadingTiles.push(
      <GridListTile key={index} data-testid='loading-tile'>
        <LoadingTile />
      </GridListTile>
    )
  }

  return loadingTiles;
}

export const getStreamTiles = (streams) => (
  streams.map((stream, index) => (
    <GridListTile key={index}>  
      <StreamTile {...stream} />
    </GridListTile>
  )
))

const GameStreamGrid = ({ children }) => (
  <GridList cols={4} cellHeight={300} spacing={20}>
    {children}
  </GridList>
)

GameStreamGrid.propTypes = {
  children: node.isRequired,
}

export default GameStreamGrid;
