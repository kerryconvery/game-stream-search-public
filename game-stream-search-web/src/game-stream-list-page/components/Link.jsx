import { styled } from '@material-ui/core/styles';

const Link = styled('a')({
  textDecoration: 'none',
  color: 'inherit',
  '&:hover > button': {
    opacity: 0.7
  },
})

export default Link;