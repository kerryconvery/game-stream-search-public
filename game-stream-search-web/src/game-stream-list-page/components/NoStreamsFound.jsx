import React from 'react';
import { string } from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles(() => ({
  root: {
    textAlign: 'center',
  },
  notFoundMessage: {
    'font-family': 'Helvetica',
    color: '#231E1D',
  }
}));

const NoStreamsFound = ({ searchTerm }) => {
  const classess = useStyles();

  return  (
    <div className={classess.root} data-testid='streams-not-found'>
      <h3 className={classess.notFoundMessage}>{`No streams were found for ${searchTerm}`}</h3>
    </div>
  )
}

NoStreamsFound.propTypes = {
  searchTerm: string,
}

NoStreamsFound.defaultProps = {
  searchTerm: '',
}

export default NoStreamsFound;