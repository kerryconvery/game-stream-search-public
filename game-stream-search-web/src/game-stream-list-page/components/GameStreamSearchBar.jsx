import React from 'react';
import GameSearchInput from './GameSearchInput';

const GameStreamSearchBar = (props) => (
  <GameSearchInput {...props} />
);

GameStreamSearchBar.propTypes = {
  ...GameSearchInput.PropTypes
}

export default GameStreamSearchBar;
