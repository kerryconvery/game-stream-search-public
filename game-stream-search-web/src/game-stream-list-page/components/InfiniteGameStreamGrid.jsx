import React from 'react';
import { string, func, bool, arrayOf, shape, number } from 'prop-types';
import InfiniteScroll from 'react-infinite-scroller';
import GameStreamGrid, { getLoadingTiles, getStreamTiles } from './GameStreamGrid';

const InfiniteGameStreamGrid = ({ streams, isLoadingStreams, hasMoreStreams, loadMoreStreams, }) => (
  <InfiniteScroll
    pageStart={0}
    loadMore={loadMoreStreams}
    hasMore={hasMoreStreams}
  >
    <GameStreamGrid>
      {getStreamTiles(streams)}
      {isLoadingStreams && getLoadingTiles(6)}
    </GameStreamGrid>
  </InfiniteScroll>
)

InfiniteGameStreamGrid.propTypes = {
  streams: arrayOf(shape({
    streamerName: string.isRequired,
    streamTitle: string.isRequired,
    streamThumbnailUrl: string.isRequired,
    streamerAvatarUrl: string.isRequired,
    platformName: string.isRequired,
    isLive: bool.isRequired,
    streamUrl: string.isRequired,
    views: number.isRequired,
  })),
  hasMoreStreams: bool.isRequired,
  isLoadingStreams: bool.isRequired,
  loadMoreStreams: func.isRequired,
}

InfiniteGameStreamGrid.defaultProps = {
  streams: [],
}

export default InfiniteGameStreamGrid;