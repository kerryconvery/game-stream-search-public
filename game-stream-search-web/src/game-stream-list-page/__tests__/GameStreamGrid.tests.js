import React from 'react';
import { render } from '@testing-library/react';
import { getStreamTiles, getLoadingTiles } from '../components/GameStreamGrid';

describe('Game stream grid tile', () => {
  it('should match the snapshot', () => {
    const streams = [
      {
        streamTitle: 'test stream A',
        streamerName: 'test streamer',
        streamThumbnailUrl: 'http://test.stream.thumbnail.url',
        streamUrl: 'http://test.stream.url',
        streamerAvatarUrl: 'http://test.channel.thumbnail.url',
        platformName: 'test platform',
        views: 100
      },
      {
        streamTitle: 'test stream B',
        streamerName: 'test streamer',
        streamThumbnailUrl: 'http://test.stream.thumbnail.url',
        streamUrl: 'http://test.stream.url',
        streamerAvatarUrl: 'http://test.channel.thumbnail.url',
        platformName: 'test platform',
        views: 10
      },
    ]

    const tiles = getStreamTiles(streams);
    const { container } = render(<div>{tiles}</div>);

    expect(tiles.length).toEqual(2);
    expect(container.firstChild).toMatchSnapshot();
  })
})


describe('Loading grid tile', () => {
  it('should match the snapshot', () => {
    const tiles = getLoadingTiles(2);
    const { container } = render(<div>{tiles}</div>);

    expect(tiles.length).toEqual(2);
    expect(container.firstChild).toMatchSnapshot();
  })
})


