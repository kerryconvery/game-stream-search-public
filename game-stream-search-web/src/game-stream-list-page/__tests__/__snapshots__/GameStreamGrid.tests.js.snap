// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game stream grid tile should match the snapshot 1`] = `
<div>
  <li
    class="MuiGridListTile-root"
  >
    <div
      class="MuiGridListTile-tile"
    >
      <div
        class="makeStyles-root-1"
      >
        <a
          class="makeStyles-link-3"
          href="http://test.stream.url"
          target="_blank"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root makeStyles-playButton-7"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root makeStyles-playButtonIcon-8"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 16.5l6-4.5-6-4.5v9zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <img
            height="180"
            src="http://test.stream.thumbnail.url"
            width="320"
          />
          <div
            class="makeStyles-detailsContainer-4"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circle"
            >
              <img
                class="MuiAvatar-img"
                src="http://test.channel.thumbnail.url"
              />
            </div>
            <div
              class="makeStyles-streamDetails-5"
            >
              <h3
                class="makeStyles-streamTitle-2"
                title="test stream A"
              >
                test stream A
              </h3>
              <div
                class="makeStyles-streamSubDetails-6"
              >
                <div>
                  test streamer
                </div>
                <span>
                  test platform
                </span>
                <span>
                   • 
                </span>
                <span>
                  100 viewers
                </span>
              </div>
            </div>
          </div>
        </a>
      </div>
    </div>
  </li>
  <li
    class="MuiGridListTile-root"
  >
    <div
      class="MuiGridListTile-tile"
    >
      <div
        class="makeStyles-root-1"
      >
        <a
          class="makeStyles-link-3"
          href="http://test.stream.url"
          target="_blank"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root makeStyles-playButton-7"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root makeStyles-playButtonIcon-8"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 16.5l6-4.5-6-4.5v9zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <img
            height="180"
            src="http://test.stream.thumbnail.url"
            width="320"
          />
          <div
            class="makeStyles-detailsContainer-4"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circle"
            >
              <img
                class="MuiAvatar-img"
                src="http://test.channel.thumbnail.url"
              />
            </div>
            <div
              class="makeStyles-streamDetails-5"
            >
              <h3
                class="makeStyles-streamTitle-2"
                title="test stream B"
              >
                test stream B
              </h3>
              <div
                class="makeStyles-streamSubDetails-6"
              >
                <div>
                  test streamer
                </div>
                <span>
                  test platform
                </span>
                <span>
                   • 
                </span>
                <span>
                  10 viewers
                </span>
              </div>
            </div>
          </div>
        </a>
      </div>
    </div>
  </li>
</div>
`;

exports[`Loading grid tile should match the snapshot 1`] = `
<div>
  <li
    class="MuiGridListTile-root"
    data-testid="loading-tile"
  >
    <div
      class="MuiGridListTile-tile"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-wave"
        style="height: 60%;"
      />
      <div
        class="makeStyles-detailsContainer-9"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-circle MuiSkeleton-wave"
          style="width: 50px; height: 50px;"
        />
        <div
          class="makeStyles-details-10"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
          />
          <div
            class="makeStyles-subDetails-11"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
            />
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
            />
          </div>
        </div>
      </div>
    </div>
  </li>
  <li
    class="MuiGridListTile-root"
    data-testid="loading-tile"
  >
    <div
      class="MuiGridListTile-tile"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-wave"
        style="height: 60%;"
      />
      <div
        class="makeStyles-detailsContainer-9"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-circle MuiSkeleton-wave"
          style="width: 50px; height: 50px;"
        />
        <div
          class="makeStyles-details-10"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
          />
          <div
            class="makeStyles-subDetails-11"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
            />
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
            />
          </div>
        </div>
      </div>
    </div>
  </li>
</div>
`;
